packages/@expressive-code/plugin-shiki/src/index.ts

```diff
@@ -44,8 +44,18 @@ export function pluginShiki(): ExpressiveCodePlugin {
 				}
 
 				// Run Shiki on the code (without explanations to improve performance)
-				const tokenLines = highlighter.codeToThemedTokens(code, codeBlock.language, theme.name, { includeExplanation: false })
+				const tokenLines =
+					codeBlock.language === 'ansi'
+						? highlighter.ansiToThemedTokens(code, theme.name)
+						: highlighter.codeToThemedTokens(code, codeBlock.language, theme.name, { includeExplanation: false })
+
 				tokenLines.forEach((line, lineIndex) => {
+					if (codeBlock.language === 'ansi') {
+						// Update line to trimmed version without sequence characters before adding styles
+						const trimmedLine = line.map((token) => token.content).join('')
+						codeLines[lineIndex].editText(undefined, undefined, trimmedLine)
```

Comment:
Using `editText` like this replaces the entire line, which would remove any annotations added to this line by other plugins.

We need to be more surgically precise here and delete only the character ranges that were actually removed by the ANSI parser. I don't think that we have direct access to those ranges, so I'd suggest adding a utility function that compares an original line to its edited version (your `trimmedLine`), and that returns an array of column ranges that were removed:

```ts
/**
 * Compares a given `original` string to its `edited` version, assuming that the only kind of edits
 * allowed between them is the removal of column ranges from the original string.
 *
 * Returns an array of column ranges that were removed from the original string.
 */
function getRemovedRanges(original: string, edited: string) {
	const ranges: [start: number, end: number][] = []
	let from = -1
	let orgIdx = 0
	let edtIdx = 0

	while (orgIdx < original.length && edtIdx < edited.length) {
		if (original[orgIdx] !== edited[edtIdx]) {
			if (from === -1) from = orgIdx
			orgIdx++
		} else {
			if (from > -1) {
				ranges.push([from, orgIdx])
				from = -1
			}
			orgIdx++
			edtIdx++
		}
	}

	if (edtIdx < edited.length) throw new Error(`Edited string contains characters not present in original (${JSON.stringify({ original, edited })})`)

	if (orgIdx < original.length) ranges.push([orgIdx, original.length])

	return ranges
}
```

We could then loop through the resulting array in reverse order and call `codeLines[lineIndex].editText(start, end, '')` to replace all the removed ANSI escape sequences with an empty string.

Reply:
Good catch and thanks for the suggestion. I misread the code in `expressive-code` and assumed Shiki would always be first unless disabled, but I see you can manually configure it to appear later. Will replace the shortcut I ended up with. 🙂

Reply:
The different types of hooks are called in a predetermined order defined in `render-block.ts`. You can [see here](https://github.com/expressive-code/expressive-code/blob/b3f92af711a0614484238e354c45519baa46a4de/packages/%40expressive-code/core/src/internal/render-block.ts#L56-L61) that there are multiple hooks which will always be called before the `performSyntaxAnalysis` hook we're using in `plugin-shiki`.

Both the `preprocessMetadata` and `preprocessCode` hooks are very likely to add inline annotations to lines, so I don't want to risk overriding those. :)